core/src/main/java/com/graphhopper/storage/CHBuilder.java

```diff
@@ -0,0 +1,126 @@
+                throw new IllegalArgumentException("Invalid level for node " + a + ": " + getLevel(a) + ". The level " +
+                        "must be equal to or larger than the lower level node of the previous shortcut (node: " + prevNodeA +
+                        ", level: " + prevLevelA + ")");
+            }
+        }
+        storage.shortcutNodeBased(a, b, accessFlags, weight, skippedEdge1, skippedEdge2);
+        int scId = storage.shortcutCount - 1;
+
+        // we keep track of the last shortcut for each node (-1 if there are no shortcuts), but
+        // we do not register the edge at node b which should be the higher level node (so no need to 'see' the lower
+        // level node a)
+        storage.setEdgeRef(storage.toNodePointer(a), origEdges + scId);
+
+        if (accessFlags == PrepareEncoder.getScFwdDir()) {
+            // todonow: get rid of origEdges here?
+            setShortcutForPrepareEdge(prepareEdgeFwd, scId + origEdges);
+        } else if (accessFlags == PrepareEncoder.getScBwdDir()) {
+            setShortcutForPrepareEdge(prepareEdgeBwd, scId + origEdges);
+        } else {
+            setShortcutForPrepareEdge(prepareEdgeFwd, scId + origEdges);
+            setShortcutForPrepareEdge(prepareEdgeBwd, scId + origEdges);
+        }
+    }
+
+    /**
+     * like shortcut(), but for edge-based CH
+     *
+     * @param origFirst The first original edge that is skipped by this shortcut. For example for the following shortcut
+     *                  edge from x to y, which itself skips the shortcuts x->v and v->y the first original edge would
+     *                  be x->u: x->u->v->w->y
+     * @param origLast  like origFirst, but the last orig edge, i.e w->y in above example
+     */
+    public void addShortcutEdgeBased(int a, int b, int accessFlags, double weight, int skippedEdge1, int skippedEdge2, int origFirst, int origLast, int prepareEdgeFwd, int prepareEdgeBwd) {
+        // todo: assert storage is edge based
+        addShortcutNodeBased(a, b, accessFlags, weight, skippedEdge1, skippedEdge2, prepareEdgeFwd, prepareEdgeBwd);
+        long shortcutPointer = storage.toShortcutPointer(storage.shortcutCount - 1);
+        storage.setOrigEdges(shortcutPointer, origFirst, origLast);
+    }
+
+    public void setShortcutForPrepareEdge(int prepareEdge, int shortcut) {
+        int index = prepareEdge - origEdges;
+        if (index >= shortcutsByPrepareEdges.size())
+            shortcutsByPrepareEdges.resize(index + 1);
+        shortcutsByPrepareEdges.set(index, shortcut);
+    }
+
+    public int getShortcutForPrepareEdge(int prepareEdge) {
+        if (prepareEdge < origEdges)
+            return prepareEdge;
+        int index = prepareEdge - origEdges;
+        return shortcutsByPrepareEdges.get(index);
+    }
+
+    public void buildCH() {
+        // during contraction the skip1/2 edges of shortcuts refer to the prepare edge-ids *not* the CHGraphImpl (shortcut)
```

Comment:
RoutingCHGraphImpl?

Reply:
Yes, I thought about renaming `RoutingCHGraph` back to `CHGraph` once `CHGraph` is gone? Otherwise yes, there is still some clean up to do here.

Reply:
Hm, not sure yet, maybe do this later.., or at least not in this PR